lib/utils/hostid/hostid.go

```diff
@@ -0,0 +1,144 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package hostid
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/google/renameio/v2"
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	// FileName is the file name where the host UUID file is stored
+	FileName = "host_uuid"
+)
+
+// GetPath returns the path to the host UUID file given the data directory.
+func GetPath(dataDir string) string {
+	return filepath.Join(dataDir, FileName)
+}
+
+// ExistsLocally checks if dataDir/host_uuid file exists in local storage.
+func ExistsLocally(dataDir string) bool {
+	_, err := ReadFile(dataDir)
+	return err == nil
+}
+
+// ReadFile reads host UUID from the file in the data dir
+func ReadFile(dataDir string) (string, error) {
+	out, err := utils.ReadPath(GetPath(dataDir))
+	if err != nil {
+		if errors.Is(err, fs.ErrPermission) {
+			//do not convert to system error as this loses the ability to compare that it is a permission error
+			return "", trace.Wrap(err)
+		}
+		return "", trace.ConvertSystemError(err)
+	}
+	id := strings.TrimSpace(string(out))
+	if id == "" {
+		return "", trace.NotFound("host uuid is empty")
+	}
+	return id, nil
+}
+
+// WriteFile writes host UUID into a file
+func WriteFile(dataDir string, id string) error {
+	err := os.WriteFile(GetPath(dataDir), []byte(id), os.ModeExclusive|0400)
+	if err != nil {
+		if errors.Is(err, fs.ErrPermission) {
+			//do not convert to system error as this loses the ability to compare that it is a permission error
+			return err
+		}
+		return trace.ConvertSystemError(err)
+	}
+	return nil
+}
+
+// ReadOrCreateFile looks for a hostid file in the data dir. If present,
+// returns the UUID from it, otherwise generates one
+func ReadOrCreateFile(dataDir string) (string, error) {
+	hostUUIDFileName := GetPath(dataDir)
+	hostUUIDFileLock := hostUUIDFileName + ".lock"
+
+	const iterationLimit = 3
+	for i := 0; i < iterationLimit; i++ {
+		id, err := ReadFile(dataDir)
+		if err == nil {
+			return id, nil
+		}
+		if !trace.IsNotFound(err) {
+			return "", trace.Wrap(err)
+		}
+
+		// Checking error instead of the usual uuid.New() in case uuid generation
+		// fails due to not enough randomness. It's been known to happen happen when
+		// Teleport starts very early in the node initialization cycle and /dev/urandom
+		// isn't ready yet.
+		rawID, err := uuid.NewRandom()
+		if err != nil {
+			return "", trace.BadParameter("" +
+				"Teleport failed to generate host UUID. " +
+				"This may happen if randomness source is not fully initialized when the node is starting up. " +
+				"Please try restarting Teleport again.")
+		}
+
+		id = rawID.String()
+
+		writeFile := func() (string, error) {
```

Comment:
Something like this? And then we can do `id, err := writeFile(rawID.String())` or something along those lines.

```suggestion
		writeFile := func(potentialNewID string) (actualID string, error) {
```